### PR TITLE
improve top_names and bottom_names in pycaffe

### PR DESCRIPTION
**Pull request by nitnelave**
By rewriting these properties to proper OrderedDict, a lot of functionalities like iterations are added to the field. Plus, it gets computed only once, instead of partially at every element access.

### CODE REVIEW

File: python/caffe/pycaffe.py [hunk 1]
```diff
@@ -292,21 +292,30 @@ def _Net_batch(self, blobs):
...
+    Parameters
+    ----------
+    func: function id -> [id]
+    field: implementation field name (cache)
+
+    Returns
+    ------
+    A one-parameter function that can be set as a property.
+    """
+    def get_id_name(self):
```
Comment (longjon) [inline python/caffe/pycaffe.py]: Why not `@property` this and return a property directly? (That'll be more uniform with the rest of this file.)

**Comment by longjon** [comment]
Thanks, I asked for this here: https://github.com/BVLC/caffe/pull/2865#issuecomment-169873421 :)

Looks good, just one comment above.

**Comment by nitnelave** [comment]
Thanks for the catch, I forgot that I could decorate inner functions too.

**Comment by longjon** [comment]
Code looks good, history looks good, passing test included: merging, thanks!

**Pull request merged**